server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -4593,11 +4600,98 @@ public void syncSegmentsFromRemoteSegmentStore(boolean overrideLocal, boolean re
+        RemoteSegmentStoreDirectory remoteSegmentStoreDirectory = null;
+        if (remoteStore != null) {
+            remoteSegmentStoreDirectory = getRemoteSegmentDirectoryForShard();
+            remoteSegmentStoreDirectory.init();
+            remoteStore.incRef();
+        }
+        sourceRemoteSegmentDirectory.init();
+        Map<String, RemoteSegmentStoreDirectory.UploadedSegmentMetadata> uploadedSegments = sourceRemoteSegmentDirectory
+            .getSegmentsUploadedToRemoteStore(primaryTerm, commitGeneration);
+        final Directory storeDirectory = store.directory();
+        store.incRef();
+
+        try {
+            copySegmentFiles(storeDirectory, sourceRemoteSegmentDirectory, remoteSegmentStoreDirectory, uploadedSegments, overrideLocal);
+        } catch (IOException e) {
+            throw new IndexShardRecoveryException(shardId, "Exception while copying segment files from remote segment store", e);
+        } finally {
+            store.decRef();
+            if (remoteStore != null) {
+                remoteStore.incRef();
+            }
+        }
+    }
+
+    private void copySegmentFiles(
+        Directory storeDirectory,
+        RemoteSegmentStoreDirectory sourceRemoteDirectory,
+        RemoteSegmentStoreDirectory targetRemoteDirectory,
+        Map<String, RemoteSegmentStoreDirectory.UploadedSegmentMetadata> uploadedSegments,
+        boolean overrideLocal
+    ) throws IOException {
+        List<String> downloadedSegments = new ArrayList<>();
+        List<String> skippedSegments = new ArrayList<>();
+
+        try {
+            Set<String> localSegmentFiles = Sets.newHashSet(storeDirectory.listAll());
+            for (String file : uploadedSegments.keySet()) {
+                long checksum = Long.parseLong(uploadedSegments.get(file).getChecksum());
+
+                if (overrideLocal || localDirectoryContains(storeDirectory, file, checksum) == false) {
+                    if (localSegmentFiles.contains(file)) {
+                        storeDirectory.deleteFile(file);
+                    }
+                    if (targetRemoteDirectory != null) {
+                        String uploadedFileName = uploadedSegments.get(file).toString().split("::")[1];
+                        logger.debug(
+                            "local file is "
+                                + file
+                                + "uploaded file name is "
+                                + uploadedFileName
+                                + " and file is "
+                                + uploadedSegments.get(file).toString()
+                        );
+                        targetRemoteDirectory.copyFrom(
+                            sourceRemoteDirectory,
+                            file,
+                            file,
+                            IOContext.DEFAULT,
+                            false,
+                            uploadedSegments.get(file).getChecksum()
+                        );
```

Comment:
this would be slower than copying it from `storeDirectory` as it reading from remote twice . Also I would be more comfortable if we are uploading directly from store directory

Reply:
I would prefer if we call `refresh`  internally before finishing the restore .

Reply:
yeah.. refresh listener today reads the latest segment infos file and then syncs that data, however today we do not commit before bringing the shard into STARTED state.

Reply:
okay , in that case lets copy data not from remote , but from local store. This is to make sure we adhere to the current direction of data transfer - local -> remote and remote -> local and not introduce more modes which is remote-> remote . Else for data integrity and encryption , we will have one more use case to support.

Reply:
ack